보안 패스 한 번 더 돌려서 발견한 추가 이슈 수정 (커밋 \`03628f7\`).

## 추가 발견 — cause chain 우회 누출

이전 커밋 \`e71652b\`에서 \`ExternalApiException\`의 message는 sanitize했지만, cause로 보존한 원본 \`ResourceAccessException\`은 여전히 자기 message에 URL+apiKey를 통째로 들고 있었음.

후속 \`GlobalExceptionHandler\`가 일반 패턴(\`log.error(msg, ex)\`)으로 로깅하면 stack trace에:
\`\`\`
Caused by: org.springframework.web.client.ResourceAccessException:
  I/O error on GET request for \"https://...&apiKey=...\"
\`\`\`
이대로 출력 → DevTools/Sentry/CloudWatch 등에 누출.

## 수정

- 모든 catch 블록에서 \`cause\` 인자를 \`null\`로 변경
- 진단 정보(\`httpStatus\`, \`Type\`, \`source\`, \`causeName\` String)는 ExternalApiException 자체 필드/메시지로 충분히 보존
- stack trace는 ExternalApiException 시점부터만 보이지만 캡스톤 디버깅엔 충분

## 추가 (일관성)

- KakaoLocalClient에 빈 body 가드 추가 — OdsayClient와 동일 패턴

머지 전 잡아야 할 보안 이슈는 이제 없을 거 같음. 추가 리뷰 부탁드립니다.

--- a/backend/src/main/java/com/todayway/backend/external/ExternalApiConfig.java
+++ b/backend/src/main/java/com/todayway/backend/external/ExternalApiConfig.java
@@ -1,0 +1,11 @@
+package com.todayway.backend.external;
+
+import com.todayway.backend.external.kakao.KakaoLocalProperties;
+import com.todayway.backend.external.odsay.OdsayProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties({OdsayProperties.class, KakaoLocalProperties.class})
+public class ExternalApiConfig {
+}

--- a/backend/src/main/java/com/todayway/backend/external/ExternalApiException.java
+++ b/backend/src/main/java/com/todayway/backend/external/ExternalApiException.java
@@ -1,0 +1,24 @@
+package com.todayway.backend.external;
+
+import lombok.Getter;
+
+@Getter
+public class ExternalApiException extends RuntimeException {
+
+    public enum Type {
+        TIMEOUT,
+        API_FAILED,
+        NETWORK
+    }
+
+    private final String source;
+    private final Type type;
+    private final Integer httpStatus;
+
+    public ExternalApiException(String source, Type type, Integer httpStatus, String message, Throwable cause) {
+        super(message, cause);
+        this.source = source;
+        this.type = type;
+        this.httpStatus = httpStatus;
+    }
+}

--- a/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
@@ -59,10 +59,12 @@ public class KakaoLocalClient {
             ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
                     ? ExternalApiException.Type.TIMEOUT
                     : ExternalApiException.Type.NETWORK;
-            throw new ExternalApiException(SOURCE, type, null, "Kakao Local 통신 실패: " + e.getMessage(), e);
+            // 일관성: e.getMessage()에 URL 박힘. ODsay와 동일 패턴으로 cause 클래스명만.
+            String causeName = rootCause != null ? rootCause.getClass().getSimpleName() : "ResourceAccessException";
+            throw new ExternalApiException(SOURCE, type, null, "Kakao Local 통신 실패 (" + causeName + ")", e);
         } catch (RestClientException e) {
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
-                    "Kakao Local 호출 중 예외: " + e.getMessage(), e);
+                    "Kakao Local 호출 중 예외 (" + e.getClass().getSimpleName() + ")", e);
         }
     }
 }

--- a/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
@@ -44,27 +44,35 @@ public class KakaoLocalClient {
      */
     public KakaoLocalSearchResponse searchKeyword(String query) {
         try {
-            return restClient.get()
+            KakaoLocalSearchResponse res = restClient.get()
                     .uri(uri -> uri.path("/search/keyword.json")
                             .queryParam("query", query)
                             .build())
                     .retrieve()
                     .body(KakaoLocalSearchResponse.class);
+
+            // Kakao가 빈 응답을 줄 때 후속 호출자의 NPE 방지. (OdsayClient와 일관)
+            if (res == null) {
+                throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
+                        null, "Kakao Local 응답 본문이 비어있음", null);
+            }
+            return res;
         } catch (RestClientResponseException e) {
+            // 보안: cause는 응답 본문 일부 포함 가능 → 보존 X.
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
-                    e.getStatusCode().value(), "Kakao Local 호출 실패: HTTP " + e.getStatusCode(), e);
+                    e.getStatusCode().value(), "Kakao Local 호출 실패: HTTP " + e.getStatusCode(), null);
         } catch (ResourceAccessException e) {
             // cause chain 끝까지 가서 timeout 판정 (factory 교체 시 안전).
             Throwable rootCause = NestedExceptionUtils.getMostSpecificCause(e);
             ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
                     ? ExternalApiException.Type.TIMEOUT
                     : ExternalApiException.Type.NETWORK;
-            // 일관성: e.getMessage()에 URL 박힘. ODsay와 동일 패턴으로 cause 클래스명만.
+            // 보안: ODsay와 동일 패턴. cause는 URL을 메시지에 들고 있어 보존 시 stack trace 우회 누출 위험.
             String causeName = rootCause != null ? rootCause.getClass().getSimpleName() : "ResourceAccessException";
-            throw new ExternalApiException(SOURCE, type, null, "Kakao Local 통신 실패 (" + causeName + ")", e);
+            throw new ExternalApiException(SOURCE, type, null, "Kakao Local 통신 실패 (" + causeName + ")", null);
         } catch (RestClientException e) {
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
-                    "Kakao Local 호출 중 예외 (" + e.getClass().getSimpleName() + ")", e);
+                    "Kakao Local 호출 중 예외 (" + e.getClass().getSimpleName() + ")", null);
         }
     }
 }

--- a/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalClient.java
@@ -1,0 +1,68 @@
+package com.todayway.backend.external.kakao;
+
+import com.todayway.backend.external.ExternalApiException;
+import com.todayway.backend.external.kakao.dto.KakaoLocalSearchResponse;
+import org.springframework.core.NestedExceptionUtils;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.RestClient;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+
+/**
+ * 카카오 로컬 키워드 검색 API 클라이언트.
+ * 명세 §8.1: /geocode 엔드포인트의 외부 호출자.
+ */
+@Component
+public class KakaoLocalClient {
+
+    private static final String SOURCE = "KAKAO_LOCAL";
+
+    private final RestClient restClient;
+
+    public KakaoLocalClient(KakaoLocalProperties properties) {
+        SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
+        factory.setConnectTimeout(Duration.ofSeconds(properties.getTimeoutSeconds()));
+        factory.setReadTimeout(Duration.ofSeconds(properties.getTimeoutSeconds()));
+
+        this.restClient = RestClient.builder()
+                .baseUrl(properties.getBaseUrl())
+                .requestFactory(factory)
+                .defaultHeader("Authorization", "KakaoAK " + properties.getApiKey())
+                .build();
+    }
+
+    /**
+     * 키워드로 장소 검색.
+     *
+     * @param query 주소 또는 장소명 (예: "국민대학교")
+     * @return 검색 결과 (matched 여부는 documents.isEmpty()로 판정)
+     */
+    public KakaoLocalSearchResponse searchKeyword(String query) {
+        try {
+            return restClient.get()
+                    .uri(uri -> uri.path("/search/keyword.json")
+                            .queryParam("query", query)
+                            .build())
+                    .retrieve()
+                    .body(KakaoLocalSearchResponse.class);
+        } catch (RestClientResponseException e) {
+            throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
+                    e.getStatusCode().value(), "Kakao Local 호출 실패: HTTP " + e.getStatusCode(), e);
+        } catch (ResourceAccessException e) {
+            // cause chain 끝까지 가서 timeout 판정 (factory 교체 시 안전).
+            Throwable rootCause = NestedExceptionUtils.getMostSpecificCause(e);
+            ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
+                    ? ExternalApiException.Type.TIMEOUT
+                    : ExternalApiException.Type.NETWORK;
+            throw new ExternalApiException(SOURCE, type, null, "Kakao Local 통신 실패: " + e.getMessage(), e);
+        } catch (RestClientException e) {
+            throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
+                    "Kakao Local 호출 중 예외: " + e.getMessage(), e);
+        }
+    }
+}

--- a/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalProperties.java
+++ b/backend/src/main/java/com/todayway/backend/external/kakao/KakaoLocalProperties.java
@@ -1,0 +1,14 @@
+package com.todayway.backend.external.kakao;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@Getter
+@Setter
+@ConfigurationProperties(prefix = "kakao-local")
+public class KakaoLocalProperties {
+    private String apiKey;
+    private String baseUrl;
+    private int timeoutSeconds = 5;
+}

--- a/backend/src/main/java/com/todayway/backend/external/kakao/dto/KakaoLocalSearchResponse.java
+++ b/backend/src/main/java/com/todayway/backend/external/kakao/dto/KakaoLocalSearchResponse.java
@@ -1,0 +1,33 @@
+package com.todayway.backend.external.kakao.dto;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+/**
+ * 카카오 로컬 키워드 검색 API 응답 (https://dapi.kakao.com/v2/local/search/keyword.json).
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record KakaoLocalSearchResponse(
+        Meta meta,
+        List<Document> documents
+) {
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public record Meta(
+            @JsonProperty("total_count") int totalCount,
+            @JsonProperty("pageable_count") int pageableCount,
+            @JsonProperty("is_end") boolean isEnd
+    ) {}
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public record Document(
+            String id,
+            @JsonProperty("place_name") String placeName,
+            @JsonProperty("address_name") String addressName,
+            @JsonProperty("road_address_name") String roadAddressName,
+            @JsonProperty("category_group_code") String categoryGroupCode,
+            String x,
+            String y
+    ) {}
+}

--- a/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
@@ -71,10 +71,13 @@ public class OdsayClient {
             ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
                     ? ExternalApiException.Type.TIMEOUT
                     : ExternalApiException.Type.NETWORK;
-            throw new ExternalApiException(SOURCE, type, null, "ODsay 통신 실패: " + e.getMessage(), e);
+            // 보안: e.getMessage()에 URL+apiKey 통째로 박힘 ("I/O error on GET request for ...&apiKey=..."). cause 클래스명만 노출.
+            String causeName = rootCause != null ? rootCause.getClass().getSimpleName() : "ResourceAccessException";
+            throw new ExternalApiException(SOURCE, type, null, "ODsay 통신 실패 (" + causeName + ")", e);
         } catch (RestClientException e) {
+            // 보안: 동일 사유로 메시지에 URL 박힐 가능. 클래스명만.
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
-                    "ODsay 호출 중 예외: " + e.getMessage(), e);
+                    "ODsay 호출 중 예외 (" + e.getClass().getSimpleName() + ")", e);
         }
     }
 }

--- a/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
@@ -62,8 +62,10 @@ public class OdsayClient {
             }
             return body;
         } catch (RestClientResponseException e) {
+            // 보안: cause로 보존하지 않음. e는 응답 본문 일부를 메시지에 담을 수 있어 우회 누출 통로.
+            // 진단에 필요한 정보(httpStatus, type, source)는 ExternalApiException 필드에 담음.
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
-                    e.getStatusCode().value(), "ODsay 호출 실패: HTTP " + e.getStatusCode(), e);
+                    e.getStatusCode().value(), "ODsay 호출 실패: HTTP " + e.getStatusCode(), null);
         } catch (ResourceAccessException e) {
             // cause chain 끝까지 가서 timeout 판정 (SimpleClientHttpRequestFactory는 SocketTimeoutException으로
             // connect/read timeout 모두 던지지만, 향후 factory 교체 시 안전하도록 root cause 검사).
@@ -71,13 +73,14 @@ public class OdsayClient {
             ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
                     ? ExternalApiException.Type.TIMEOUT
                     : ExternalApiException.Type.NETWORK;
-            // 보안: e.getMessage()에 URL+apiKey 통째로 박힘 ("I/O error on GET request for ...&apiKey=..."). cause 클래스명만 노출.
+            // 보안: e.getMessage()는 URL+apiKey를 통째로 포함하고, cause로 보존하면 stack trace에서 우회 누출.
+            // 메시지엔 cause 클래스명만 + cause는 null로 끊음. 진단 정보는 type/source 필드로 충분.
             String causeName = rootCause != null ? rootCause.getClass().getSimpleName() : "ResourceAccessException";
-            throw new ExternalApiException(SOURCE, type, null, "ODsay 통신 실패 (" + causeName + ")", e);
+            throw new ExternalApiException(SOURCE, type, null, "ODsay 통신 실패 (" + causeName + ")", null);
         } catch (RestClientException e) {
-            // 보안: 동일 사유로 메시지에 URL 박힐 가능. 클래스명만.
+            // 보안: 동일 사유로 cause 보존 X.
             throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
-                    "ODsay 호출 중 예외 (" + e.getClass().getSimpleName() + ")", e);
+                    "ODsay 호출 중 예외 (" + e.getClass().getSimpleName() + ")", null);
         }
     }
 }

--- a/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
+++ b/backend/src/main/java/com/todayway/backend/external/odsay/OdsayClient.java
@@ -1,0 +1,80 @@
+package com.todayway.backend.external.odsay;
+
+import com.todayway.backend.external.ExternalApiException;
+import org.springframework.core.NestedExceptionUtils;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.ResourceAccessException;
+import org.springframework.web.client.RestClient;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+
+/**
+ * ODsay 대중교통 길찾기 API 클라이언트.
+ * 명세 §5.1: 응답을 raw JSON 그대로 schedule.route_summary_json에 저장한다.
+ */
+@Component
+public class OdsayClient {
+
+    private static final String SOURCE = "ODSAY";
+
+    private final OdsayProperties properties;
+    private final RestClient restClient;
+
+    public OdsayClient(OdsayProperties properties) {
+        this.properties = properties;
+
+        SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
+        factory.setConnectTimeout(Duration.ofSeconds(properties.getTimeoutSeconds()));
+        factory.setReadTimeout(Duration.ofSeconds(properties.getTimeoutSeconds()));
+
+        this.restClient = RestClient.builder()
+                .baseUrl(properties.getBaseUrl())
+                .requestFactory(factory)
+                .build();
+    }
+
+    /**
+     * 대중교통 경로 검색.
+     *
+     * @param sx 출발지 경도 (longitude)
+     * @param sy 출발지 위도 (latitude)
+     * @param ex 도착지 경도
+     * @param ey 도착지 위도
+     * @return ODsay 응답 raw JSON 문자열 (route_summary_json에 그대로 저장)
+     */
+    public String searchPubTransPathT(double sx, double sy, double ex, double ey) {
+        try {
+            // URI 템플릿 변수 확장으로 query value를 strict encoding (apiKey에 '+', '/' 포함되어 필요).
+            String body = restClient.get()
+                    .uri("/searchPubTransPathT?SX={sx}&SY={sy}&EX={ex}&EY={ey}&apiKey={apiKey}",
+                            sx, sy, ex, ey, properties.getApiKey())
+                    .retrieve()
+                    .body(String.class);
+
+            // ODsay가 빈 응답을 줄 때 후속 JSON 파싱 단계의 NPE 방지.
+            if (body == null || body.isBlank()) {
+                throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
+                        null, "ODsay 응답 본문이 비어있음", null);
+            }
+            return body;
+        } catch (RestClientResponseException e) {
+            throw new ExternalApiException(SOURCE, ExternalApiException.Type.API_FAILED,
+                    e.getStatusCode().value(), "ODsay 호출 실패: HTTP " + e.getStatusCode(), e);
+        } catch (ResourceAccessException e) {
+            // cause chain 끝까지 가서 timeout 판정 (SimpleClientHttpRequestFactory는 SocketTimeoutException으로
+            // connect/read timeout 모두 던지지만, 향후 factory 교체 시 안전하도록 root cause 검사).
+            Throwable rootCause = NestedExceptionUtils.getMostSpecificCause(e);
+            ExternalApiException.Type type = rootCause instanceof SocketTimeoutException
+                    ? ExternalApiException.Type.TIMEOUT
+                    : ExternalApiException.Type.NETWORK;
+            throw new ExternalApiException(SOURCE, type, null, "ODsay 통신 실패: " + e.getMessage(), e);
+        } catch (RestClientException e) {
+            throw new ExternalApiException(SOURCE, ExternalApiException.Type.NETWORK, null,
+                    "ODsay 호출 중 예외: " + e.getMessage(), e);
+        }
+    }
+}

--- a/backend/src/main/java/com/todayway/backend/external/odsay/OdsayProperties.java
+++ b/backend/src/main/java/com/todayway/backend/external/odsay/OdsayProperties.java
@@ -1,0 +1,14 @@
+package com.todayway.backend.external.odsay;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@Getter
+@Setter
+@ConfigurationProperties(prefix = "odsay")
+public class OdsayProperties {
+    private String apiKey;
+    private String baseUrl;
+    private int timeoutSeconds = 5;
+}

--- a/backend/src/main/resources/application.yml
+++ b/backend/src/main/resources/application.yml
@@ -51,6 +51,7 @@ odsay:
 kakao-local:
   api-key: ${KAKAO_LOCAL_API_KEY:}
   base-url: https://dapi.kakao.com/v2/local
+  timeout-seconds: 5
 
 vapid:
   public-key: ${VAPID_PUBLIC_KEY:}

